cumulus/client/cirrus-executor/src/bundle_processor.rs

```diff
@@ -626,32 +652,13 @@ where
     }
 
     fn submit_fraud_proof(&self, fraud_proof: FraudProof) {
-        let primary_chain_client = self.primary_chain_client.clone();
-        // TODO: No backpressure
-        self.spawner.spawn_blocking(
-            "cirrus-submit-fraud-proof",
-            None,
-            async move {
-                tracing::debug!(
-                    target: LOG_TARGET,
-                    "Submitting fraud proof in a background task..."
-                );
-                if let Err(error) = primary_chain_client
-                    .runtime_api()
-                    .submit_fraud_proof_unsigned(
-                        &BlockId::Hash(primary_chain_client.info().best_hash),
-                        fraud_proof,
-                    )
-                {
-                    tracing::error!(
-                        target: LOG_TARGET,
-                        error = ?error,
-                        "Failed to submit fraud proof"
-                    );
-                }
-            }
-            .boxed(),
-        );
+        if let Err(msg) = self
+            .unsigned_submitter
+            .try_submit(UnsignedMessage::Fraud(fraud_proof))
+        {
+            let mut pending_unsigned = self.pending_unsigned.lock();
+            pending_unsigned.replace(msg);
```

Comment:
Hm... but this will just drop existing pending extrinsic (if there is any), is that okay? I was expecting that the pipeline will pause until sending is possible again.

Reply:
Only when no pending extrinsic can this line be reached, at the beginning of processing the bundles/gossiped messages, it will first check the pending extrinsic and try to resend, only continue the processing if the resend is okay when there is a pending extrinsic, e.g., https://github.com/subspace/subspace/pull/695/files#diff-de30ccb67696749ca21e43ded6b44ae00cd2822d3137191c39d830d80c2f81f7R534-R550. So the pipeline will pause if the message buffer is full as you expect.

Reply:
Logic is too complex, I don't see how `on_bundle` there has anything to do with `submit_fraud_proof` method that is called from a completely different place. Even if that is correct, it is extremely not obvious to me.

Reply:
I don't see any incorrectness in it, you need to check whether the data producer should continue the work at the beginning before it proceeds if you want to pause it when there are too many of them, `on_bundle` is one message source of them.

Replacing `pending_unsigned()` by checking if the channel still has some capacity should make it more obvious, and resolves the comment below.